multicluster/controllers/multicluster/memberclusterannounce_controller.go

```diff
@@ -138,10 +142,41 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
 				}
 			}
 		}
-		// If err != nil, probably ClusterClaims were deleted during the processing of MemberClusterAnnounce.
-		// Nothing to handle in this case and MemberClusterAnnounce will also be deleted soon.
-		// TODO: Add ClusterClaim webhook to make sure it cannot be deleted while ClusterSet is present.
 	}
+	// If err != nil, probably ClusterClaims were deleted during the processing of MemberClusterAnnounce.
+	// Nothing to handle in this case and MemberClusterAnnounce will also be deleted soon.
+	// TODO: Add ClusterClaim webhook to make sure it cannot be deleted while ClusterSet is present.
+
+	finalizer := fmt.Sprintf("%s/%s", MemberClusterAnnounceFinalier, memberAnnounce.ClusterID)
+	if memberAnnounce.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(memberAnnounce, finalizer) {
+			return ctrl.Result{}, nil
+		}
+		klog.InfoS("Adding member cluster to ClusterSet", "Cluster", memberAnnounce.ClusterID)
+		if err := r.addMemberToClusterSet(memberAnnounce); err != nil {
+			klog.ErrorS(err, "Failed to add member cluster to ClusterSet", "Cluster", memberAnnounce.ClusterID)
+			return ctrl.Result{}, err
+		}
+		klog.InfoS("Adding finalizer to MemberClusterAnnounce", "MemberClusterAnnounce", memberAnnounce.Name)
+		controllerutil.AddFinalizer(memberAnnounce, finalizer)
+		if err := r.Update(context.TODO(), memberAnnounce); err != nil {
+			klog.ErrorS(err, "Failed to update MemberClusterAnnounce", "MemberClusterAnnounce", memberAnnounce.Name)
+			return ctrl.Result{}, err
+		}
+
+	} else {
+		klog.InfoS("Removing member cluster from the ClusterSet", "Cluster", memberAnnounce.ClusterID)
```

Comment:
```suggestion
		klog.InfoS("Removing member cluster from the ClusterSet", "cluster", memberAnnounce.ClusterID)
```
please check other occurrences.

Reply:
Could you move deletion logic before update/create actions? Usually Reconciler will check Finalizer and `!memberAnnounce.DeletionTimestamp.IsZero()` handle deletion first, then you can return earlier to reduce unnecessary indent.

Reply:
Move the deletion logic before the update actions.